Support SSH to neighbor VMs using test server as proxy

wangxin opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] New Test case
    - [ ] Skipped for non-supported platforms
- [ ] Test case improvement


### Back port request
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411
- [ ] 202505

### Approach
#### What is the motivation for this PR?

**Problem**

Some test scripts require SSH access to neighbor VMs to perform operations. Currently, this requires direct network connectivity between the sonic-mgmt container and the neighbor VMs' management ports.

In lab environments, the sonic-mgmt container may not run on the same test server hosting the neighbor VMs. When neighbor VMs are assigned private IP addresses accessible only from the test server itself, direct SSH connections from the sonic-mgmt container fail.

**Solution**

This change enables SSH access to neighbor VMs through the test server as a proxy. Since the sonic-mgmt container can access the test server's management interface, we leverage SSH's built-in proxy capability to route connections through the test server.

#### How did you do it?

The SSH key based authentication from sonic-mgmt container to test server is mandatory. The underlying ansible SSH plugins are not able to provide credential to login the proxy.

1. **SSH key generation in testbed-cli.sh**: Updated `testbed-cli.sh` to ensure SSH public and private keys are always present in the sonic-mgmt container. While `setup-container.sh` already handles this during container creation, this change ensures key availability when containers are created through other means. These keys are required for key-based authentication between the sonic-mgmt container and test server.

2. **SSH key deployment**: Modified `ansible/roles/vm_set/tasks/main.yml` to automatically add the sonic-mgmt container's SSH public key to the test server's `~/.ssh/authorized_keys` file. This enables key-based SSH authentication and is executed during `deploy-topo` or `restart-ptf` operations.

3. **conftest enhancement**: Enhanced `tests/conftest.py` to support scenarios where the testbed is deployed in one sonic-mgmt container but tests run in another. The SSH key of the sonic-mgmt container running tests may not be added to the test server's `~/.ssh/authorized_keys` file yet. The pytest code now ensures SSH key-based authentication is enabled from the test execution container to the test server.

4. **nbrhosts fixture enhancement**: Updated the `nbrhosts` fixture to retrieve the test server's username and IP address, then pass them to the neighbor VM host constructor via the `ssh_proxy` argument. The argument format is a dictionary: `ssh_proxy={"proxy_user": "username", "proxy_host": "10.0.10.1"}`.

5. **Neighbor VM host constructors**: Enhanced neighbor VM host constructors to accept the `ssh_proxy` argument and construct appropriate Ansible variables for the underlying SSH plugin. This enables the Ansible SSH plugin to establish connections to neighbor VMs through the test server proxy.

The implementation supports two Ansible SSH plugin types used by the pytest code:

- **Regular SSH plugin**: Configure the `ansible_ssh_extra_args` variable to pass SSH proxy information. The SSH plugin will combine the `ssh_args` in `ansible/ansible.cfg` file and this `ansible_ssh_extra_args` to construct SSH command for connecting to neighbor VMs. 
``` 
evars = { 'ansible_ssh_extra_args': '-o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o ProxyCommand="ssh -W %h:%p {proxy_user}@{proxy_host}"' } 
self.host.options['variable_manager'].extra_vars.update(evars) 
```

- **Network CLI connection**: Set the `ansible_paramiko_proxy_command` variable: 
``` 
evars['ansible_paramiko_proxy_command'] = 'ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -W %h:%p {proxy_user}@{proxy_host}' 
self.host.options['variable_manager'].extra_vars.update(evars) 
```

For test server in the `veos` or `veos_vtb` inventory file, added support of an optional host variable `ansible_ssh_proxy_host`. Its value should be either IPV4 or IPv6 of the test server's IP address. The reason is that some test server's IPv4 can only be used as SSH proxy and other test server's IPv6 can only be used as SSH proxy. There is no way to automatically figure out. So added this argument to manually specify which IP can be used for proxy purpose. When this variable is not provided, by default the ansible_host will be used. This new variable is only used by the pytest code, not used by the playbook yet.

#### How did you verify/test it?
Tested run `bgp/test_bgp_gr_helper.py` from a remote sonic-mgmt container only has direct access to the test server.

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

lolyu reviewed (commented):
if a user wants to use this mechanism, the testbed needs to be redeployed, right?
if this is the case, users that wish to debug from their own sonic-mgmt container need to redeploy the testbed to make this work.
can we have a `testbed-cli.sh` function to setup the proxy only without deployment?

lolyu reviewed (dismissed):
LGTM

wangxin committed:
Support SSH to neighbor VMs using test server as proxy

Some test scripts require SSH access to neighbor VMs to perform operations. Currently, this requires direct network connectivity between the sonic-mgmt container and the neighbor VMs' management ports.

In lab environments, the sonic-mgmt container may not run on the same test server hosting the neighbor VMs. When neighbor VMs are assigned private IP addresses accessible only from the test server itself, direct SSH connections from the sonic-mgmt container fail.

This change enables SSH access to neighbor VMs through the test server as a proxy. Since the sonic-mgmt container can access the test server's management interface, we leverage SSH's built-in proxy capability to route connections through the test server.

1. **SSH key generation in testbed-cli.sh**: Updated `testbed-cli.sh` to ensure SSH public and private keys are always present in the sonic-mgmt container. While `setup-container.sh` already handles this during container creation, this change ensures key availability when containers are created through other means. These keys are required for key-based authentication between the sonic-mgmt container and test server.

2. **SSH key deployment**: Modified `ansible/roles/vm_set/tasks/main.yml` to automatically add the sonic-mgmt container's SSH public key to the test server's `~/.ssh/authorized_keys` file. This enables key-based SSH authentication and is executed during `deploy-topo` or `restart-ptf` operations.

3. **conftest enhancement**: Enhanced `tests/conftest.py` to support scenarios where the testbed is deployed in one sonic-mgmt container but tests run in another. The SSH key of the sonic-mgmt container running tests may not be added to the test server's `~/.ssh/authorized_keys` file yet. The pytest code now ensures SSH key-based authentication is enabled from the test execution container to the test server.

4. **nbrhosts fixture enhancement**: Updated the `nbrhosts` fixture to retrieve the test server's username and IP address, then pass them to the neighbor VM host constructor via the `ssh_proxy` argument. The argument format is a dictionary: `ssh_proxy={"proxy_user": "username", "proxy_host": "10.0.10.1"}`.

5. **Neighbor VM host constructors**: Enhanced neighbor VM host constructors to accept the `ssh_proxy` argument and construct appropriate Ansible variables for the underlying SSH plugin. This enables the Ansible SSH plugin to establish connections to neighbor VMs through the test server proxy.

The implementation supports two Ansible SSH plugin types used by the pytest code:

- **Regular SSH plugin**: Configure the `ansible_ssh_extra_args` variable to pass SSH proxy information. The SSH plugin will combine the `ssh_args` in `ansible/ansible.cfg` file and this `ansible_ssh_extra_args` to construct SSH command for connecting to neighbor VMs.
  ```
  evars = {
      'ansible_ssh_extra_args': '-o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o ProxyCommand="ssh -W %h:%p {proxy_user}@{proxy_host}"'
  }
  self.host.options['variable_manager'].extra_vars.update(evars)
  ```

- **Network CLI connection**: Set the `ansible_paramiko_proxy_command` variable:
  ```
  evars['ansible_paramiko_proxy_command'] = 'ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -W %h:%p {proxy_user}@{proxy_host}'
  self.host.options['variable_manager'].extra_vars.update(evars)
  ```

Signed-off-by: Xin Wang <xiwang5@microsoft.com>

wangxin committed:
Resolve conflicts

Signed-off-by: Xin Wang <xiwang5@microsoft.com>

wangxin commented:
/azp run

wangxin committed:
Fix pre-commit reported issue

Signed-off-by: Xin Wang <xiwang5@microsoft.com>

mssonicbld commented:
/azp run

yxieca reviewed (approved):
AI agent on behalf of Ying. Reviewed; no issues found.

wangxin commented:
There is still risk for this PR. And it is not really needed at this point. Close it for now. In the future, we can re-visit this PR if there is such need.
